content/blog/2019-12-18-novm/index.md

```diff
@@ -0,0 +1,317 @@
+accessing any element in an array requires only one or two instructions
+to add the offset of the base pointer and then load from memory.
+Similarly, locations on the program stack are computed relative to the current stack pointer.
+Neither of these "pointer arithmetic" operations would be valid if
+executed on the physical addresses.
+The latter is a useful security primitive that enables
+strong process memory isolation "for free," since there is no way for a process
+to even reference memory owned by another process
+(unless the OS maps some physical location into both address spaces).
+
+
+## The Case Against Virtual Addressing
+
+The translation of virtual addresses is accelerated by dedicated hardware
+called the [Translation Lookaside Buffer](https://en.wikipedia.org/wiki/Translation_lookaside_buffer) (TLB).
+This acts as a "translation cache" and hides most of the cost of virtual address translation,
+except for when an address is not present in the TLB.
+Missing in the TLB triggers a complex series of physical memory accesses
+called "walking the page table" and tends to be extremely expensive
+(especially if this has to be handled by software).
+
+For workloads that allocate very large amounts of memory, the TLB can't actually "reach"
+all of the necessary memory addresses, causing frequent
+[TLB misses](https://research.cs.wisc.edu/multifacet/papers/isca13_direct_segment.pdf).
+In these cases, it's not uncommon for the CPU to be running only a single application
+which would like to manage its own memory anyway; the aforementioned advantages
+of virtual addressing are significantly reduced but the cost in TLB misses can be devastating to performance.
+The other major cause of TLB misses is frequent context switching between processes,
+which typically triggers a complete flush of the TLB state. For multithreaded applications
+which rely heavily on system calls (e.g., webservers), this can incur
+[overheads of up to 20%](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/osr2007_rethinkingsoftwarestack.pdf).
+
+Furthermore, virtual addressing is not a requirement for memory security.
+There are many different proposals (and even some usable implementations)
+of _tagged memory_ architectures, where physical memory locations are associated with
+_tags_ that control how those locations can be accessed by software.
+Some examples include: the [CHERI capability architecture](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/);
+the [PUMP processor for software-defined metadata](https://dl.acm.org/citation.cfm?id=2694383);
+and the [secure information flow CPU, Hyperflow](https://dl.acm.org/citation.cfm?id=3243743).
+Instead of relying on a process' inability to address memory,
+these designs use hardware to efficiently check whether or not a memory access
+is allowed by the system's security policy. In these designs,
+the protection provided by virtual addressing is either mostly or completely redundant.
```

Comment:
FWIW, the other technique generally bandied about (aside from tagged memory) is some way of _decoupling protection from translation_. Here's [a recent paper](https://dl.acm.org/citation.cfm?id=3080209) that has that flavor. The rough idea is that the enforcement mechanism for memory protection might look different—and be quite a bit cheaper—than traditional page tables & TLBs if it didn't also have the responsibility of telling you where to find the memory.